### ISSUE TITLE
Add an improved RNG

### ISSUE THREAD

**Issue by FredrIQ** [roukaour/pokecrystal#42]
The current RNG is not sufficient for generating several random numbers in short succession as is needed with generation of 3 DVs, shininess, Gender, ability, nature. I intend to look into ANSI C's rng which I think is sufficient for implementation on GB hardware (Things like MT19937 or similar is not feasible).

The reason I want to change the RNG is because the current one is rather limited. It works for most purposes due to how it's used, but fails miserably at generating several random numbers. This is the current RNG:
There are 2 RAM addresses used, randomAdd and randomSub. These add and subtract respectively a hardware register called rDIV (itself increasing by 1 every 256 cycles) every VBlank and RNG call. This means that you can at most generate 3 numbers before every other outcome becomes predictable after each input (Players are assumed to do input in random amount of frames). This also caused RBY to only be able to generate about 10% of all random values (and it only does 2 RNG calls!) due to how it uses the RNG (it uses the same) for wild encounters. GSC addresses this particular issue I think by changing how it uses the RNG, but otherwise suffer from the same issues.

**Comment by FredrIQ** [comment]
ANSI C's choice of multipliers and add value complicates logic a bit too much to be worth it IMO, so I switched to another example listed on Wikipedia (and it's not RANDU) that greatly simplifies logic. The RNG itself is essentially finished (just 20 or so lines), I'll go ahead and replace the existing RNG with it once I've double-checked for typos or other silliness. Implementation: http://sprunge.us/ETDL

Going to be called on each normal randomness call, and on every VBlank.

**Comment by roukaour** [comment]
Looks like a variety of linear congruential generator. Xorshift128 is another option with period 2^128−1, although xor'ing 32-bit integers might be slower than yours.

It seems like the rDIV register could be useful for extra randomness—xor it with the final random byte? Also, parts of the game directly refer to hRandomSub and hRandomAdd.

**Comment by FredrIQ** [comment]
Yeah, it's LCG (The link has a minor typo but I've already addressed that).
I did consider rDIV for reseeding in specific circumstances as entropy (for example during DV+PV generation since it will call the RNG a lot) but I'm not sure how to properly apply it. I don't think applying it unconditionally is neccessary.
